src/express2.ml

```diff
@@ -0,0 +1,186 @@
+type done_
+(** abstract type which ensure middleware function must either
+    call the [next] function or one of the [send] function on the 
+    response object. 
+
+    This should be a great argument for OCaml, the type system 
+    prevents silly error which in this case would make the server hang *)
+
+(** TODO : maybe this should be a more common module like bs-error *)
+module Error = struct 
+
+  type t 
+
+  external message : t -> string option = "" 
+    [@@bs.send] [@@bs.return null_undefined_to_opt]
+  
+  external name : t -> string option = "" 
+    [@@bs.send] [@@bs.return null_undefined_to_opt]
+
+end 
+
+module Request = struct 
+
+  type t 
+
+  type params = Js_json.t Js_dict.t
+
+  external params : t -> params = "" [@@bs.get]
+end 
+
+module Response = struct 
+
+  type t 
+
+  external sendString : t -> string -> done_ = "send" [@@bs.send] 
+end
+
+module Next = struct 
+
+  type t = unit -> done_ [@bs]
+
+  let to_router : t -> done_ [@bs] = fun [@bs] t ->
+    (* This is the only way I could find to be able to write `next("router")` 
+     * as well *)
+    let module Cast = struct 
+      external arg : t -> (string -> done_ [@bs]) = "%identity"
+    end in 
+    (Cast.arg t) "router" [@bs]
+
+end 
+
+module Middleware = struct 
+
+  type next  = unit -> done_ [@bs]
+
+  type t 
+    (** Middleware abstract type which unified the various way an 
+        Express middleware can be constructed:
+        {ul
+        {- from a {b function} using the [ofFunction]}  
+        {- from a Router class}
+        {- from an App class}
+        {- from the third party middleware modules}   
+        }
+      
+        For each of the class which implements the middleware interface 
+        is JavaScript, one can use the "%identity" function to cast
+        it to this type [t] *)
+
+  type f = Request.t -> Response.t -> next -> done_
+
+  external ofF : f -> t = "%identity" 
+  (** Create a Middleware from a function *)
+
+  type errorF = Error.t -> Request.t -> Response.t -> next -> done_ 
+
+  external ofErrorF : errorF -> t = "%identity" 
+  (** Create a Middleware from an error function *)
+
+end 
+
+(** Generate the common Middleware binding function for a given 
+    type. This Functor is used for the Router and App classes. *)
+module MakeBindFunctions(T: sig type t end) = struct 
+
+  external use : T.t -> Middleware.t -> unit = "" [@@bs.send]
+  
+  external useOnPath : T.t -> string -> Middleware.t -> unit = "use" [@@bs.send]
+
+  external get :  
+    T.t -> 
+    string -> 
```

Comment:
Can we use labelled arguments here? `T.t -> ~path:string -> ~middleware:Middleware.t -> unit`